src/internal/permissions.ts

```diff
@@ -0,0 +1,429 @@
+}
+
+// Actions that can be granted on NFTs
+export enum NFTCapability {
+  transfer = "transfer",
+  mutate = "mutate",
+}
+
+export enum PermissionType {
+  FungibleAsset = "FungibleAsset",
+  Gas = "Gas",
+  NFT = "NFT",
+  NFTCollection = "Collection",
+}
+
+export enum MoveVMPermissionType {
+  FungibleAsset = "0x1::fungible_asset::WithdrawPermission",
+  
+}
+
+export interface FungibleAssetPermission {
+  type: PermissionType.FungibleAsset;
+  asset: string;
+  // Question: best type here?: number | string | bigint
+  balance: string;
+}
+
+export interface GasPermission {
+  type: PermissionType.Gas;
+  amount: number;
+}
+
+export interface NFTPermission {
+  type: PermissionType.NFT;
+  assetAddress: string;
+  capabilities: Record<NFTCapability, boolean>;
+}
+
+export interface NFTCollectionPermission {
+  type: PermissionType.NFTCollection;
+  collectionAddress: string;
+  capabilities: Record<NFTCollectionCapability, boolean>;
+}
+
+export enum NFTCollectionCapability {
+  transfer = "transfer",
+  mutate = "mutate",
+}
+
+export interface GenericPermission {
+  type: Exclude<string, PermissionType | `${PermissionType}`>;
+  [key: string]: string | Exclude<string, PermissionType | `${PermissionType}`>;
+}
+
+// revoke permission types
+export type RevokeFungibleAssetPermission = Pick<FungibleAssetPermission, "type" | "asset">;
+export type RevokeNFTPermission = Pick<NFTPermission, "type" | "assetAddress">;
+
+// factories -- on second thought, i think these are dumb
+export function FungibleAssetPermission(args: Omit<FungibleAssetPermission, "type">): FungibleAssetPermission {
+  return { type: PermissionType.FungibleAsset, ...args };
+}
+export function GasPermission(args: Omit<GasPermission, "type">): GasPermission {
+  return { type: PermissionType.Gas, ...args };
+}
+export function NFTPermission(args: Omit<NFTPermission, "type">): NFTPermission {
+  return { type: PermissionType.NFT, ...args };
+}
+export function NFTCollectionPermission(args: Omit<NFTCollectionPermission, "type">): NFTCollectionPermission {
+  return { type: PermissionType.NFTCollection, ...args };
+}
+
+// revoke factories
+export function RevokeFungibleAssetPermission(args: Omit<RevokeFungibleAssetPermission, "type">): RevokeFungibleAssetPermission {
+  return { type: PermissionType.FungibleAsset, ...args };
+}
+export function RevokeNFTPermission(args: Omit<RevokeNFTPermission, "type">): RevokeNFTPermission {
+  return { type: PermissionType.NFT, ...args };
+}
+
+// mostly for testing
+export function GenericPermission(args: GenericPermission): GenericPermission {
+  return {...args}
+}
+
+// functions
+
+export interface PermissionTemp{
+  asset: string;
+  type: string;
+  remaining: string;
+}
+export async function getPermissions<T extends PermissionType>({
+  aptosConfig,
+  primaryAccount,
+  subAccount,
+  filter
+}: {
+  aptosConfig: AptosConfig;
+  primaryAccount: SingleKeyAccount;
+  subAccount: AbstractedEd25519Account;
```

Comment:
Thinking about this for a moment, instead of `subAccount` we would want the public key of the sub account. Dapps will use this SDK as well as wallets, and wallets most likely won't be saving the private keys of the sub account.